discovery/hetzner/hetzner.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hetzner
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/hetznercloud/hcloud-go/hcloud"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	hcloudLabelID          = model.MetaLabelPrefix + "hcloud_id"
+	hcloudLabelName        = model.MetaLabelPrefix + "hcloud_name"
+	hcloudLabelImage       = model.MetaLabelPrefix + "hcloud_image"
+	hcloudLabelPublicIPv4  = model.MetaLabelPrefix + "hcloud_ipv4"
+	hcloudLabelPrivateIPv4 = model.MetaLabelPrefix + "hcloud_private_ipv4"
+	hcloudLabelNetwork     = model.MetaLabelPrefix + "hcloud_network"
+	hcloudLabelLocation    = model.MetaLabelPrefix + "hcloud_location"
+	hcloudLabelDatacenter  = model.MetaLabelPrefix + "hcloud_datacenter"
+	hcloudLabelCPUCores    = model.MetaLabelPrefix + "hcloud_cpu_cores"
+	hcloudLabelCPUType     = model.MetaLabelPrefix + "hcloud_cpu_type"
+	hcloudLabelMemory      = model.MetaLabelPrefix + "hcloud_memory_size"
+	hcloudLabelDisk        = model.MetaLabelPrefix + "hcloud_disk_size"
+	hcloudLabelType        = model.MetaLabelPrefix + "hcloud_server_type"
+)
+
+// DefaultSDConfig is the default Hetzner SD configuration.
+var DefaultSDConfig = SDConfig{
+	Port:            80,
+	RefreshInterval: model.Duration(60 * time.Second),
+	HcloudSDConfig: HcloudSDConfig{
+		Network:  "",
+		Endpoint: hcloud.Endpoint,
+	},
+}
+
+// SDConfig is the configuration for Hetzner based service discovery.
+type SDConfig struct {
+	HTTPClientConfig config_util.HTTPClientConfig `yaml:",inline"`
+
+	RefreshInterval model.Duration `yaml:"refresh_interval"`
+	Port            int            `yaml:"port"`
+	HcloudSDConfig  HcloudSDConfig `yaml:"hcloud"`
```

Comment:
Why not put these values in directly?

Reply:
I wanted to separate them because they won't be valid for dedicated servers and I wanted to avoid having something like `hcloud_network` `robot_vswitch` and instead have a clean object structure.

Reply:
Let's worry about that if/when it happens, and keep things simple for now. We generally don't nest for this sort of thing.